File: PMMP/src/jkorn/pvpcore/utils/PvPCKnockback.php
<?php
/**
 * Created by PhpStorm.
 * User: jkorn2324
 * Date: 2020-06-13
 * Time: 18:02
 */

declare(strict_types=1);

namespace jkorn\pvpcore\utils;


class PvPCKnockback implements IExportedValue
{

    /** @var float */
    private $xKB, $yKB;

    /** @var int */
    private $speed;

    public function __construct(float $x = 0.4, float $y = 0.4, int $speed = 10)
    {
        $this->xKB = $x;
        $this->yKB = $y;

        $this->speed = $speed;
    }

    /**
     * @param string $key
     * @param $value
     *
     * Updates the knockback value based on the key.
     */
    public function update(string $key, $value): void
    {
        switch ($key)
        {
            case Utils::X_KB:
                $this->xKB = (float)$value;
                break;
            case Utils::Y_KB:
                $this->yKB = (float)$value;
                break;
            case Utils::SPEED_KB:
                $this->speed = (int)$value;
                break;
        }
    }

    /**
     * @return float
     *
     * Gets the horizontal kb value.
     */
    public function getXZKb(): float
    {
        return $this->xKB;
    }

    /**
     * @return float
     *
     * Gets the vertical kb value.
     */
    public function getYKb(): float
    {
        return $this->yKB;
    }

    /**
     * @return int
     *
     * Gets the speed value.
     */
    public function getSpeed(): int
    {
        return $this->speed;
    }

    /**
     * @return array
     *
     * Exports the value to an array.
     */
    public function toArray(): array
    {
        return [
            Utils::X_KB => $this->xKB,
            Utils::Y_KB => $this->yKB,
            Utils::SPEED_KB => $this->speed
        ];
    }

    /**
     * @param $object
     * @return bool
     *
     * Determines if the objects are equivalent.
     */
    public function equals($object): bool
    {
        if($object instanceof PvPCKnockback)
        {
            return $object->xKB === $this->xKB
                && $object->yKB === $this->yKB
                && $object->speed === $this->speed;
        }

        return false;
    }

    /**
     * @param $data
     * @return PvPCKnockback|null
     *
     * Decodes the knockback values.
     */
    public static function decode($data): ?PvPCKnockback
    {
        if(isset($data[$x = Utils::X_KB], $data[$y = Utils::Y_KB], $data[$speed = Utils::SPEED_KB]))
        {
            return new PvPCKnockback(
                (float)$data[$x],
                (float)$data[$y],
                (int)$data[$speed]
            );
        }

        return null;
    }
}

File: Nukkit/src/pvpcore/utils/PvPCKnockback.java
package pvpcore.utils;

import org.json.simple.JSONObject;

import java.util.HashMap;
import java.util.Map;

public class PvPCKnockback implements IExportedObject {

    // Declares the constant variables.
    public static final String HORIZONTAL_KB = "kb-x";
    public static final String VERTICAL_KB = "kb-y";
    public static final String SPEED_KB = "kb-speed";

    /** The horizontal & vertical KB values. */
    private float horizontalKB, verticalKB;
    /** The attack delay value. */
    private int attackDelay;

    /**
     * The default PvPCKnockback constructor.
     */
    public PvPCKnockback()
    {
        this.horizontalKB = 0.4f;
        this.verticalKB = 0.4f;
        this.attackDelay = 10;
    }

    /**
     * The PvPCKnockback constructor.
     * @param xzKB - The horizontal input KB.
     * @param yKB - The vertical input KB.
     * @param attackDelay - The input attack delay.
     */
    public PvPCKnockback(float xzKB, float yKB, int attackDelay)
    {
        this.horizontalKB = xzKB;
        this.verticalKB = yKB;
        this.attackDelay = attackDelay;
    }

    /**
     * Gets the horizontal KB.
     * @return - The horizontal KB.
     */
    public float getHorizontalKB()
    {
        return this.horizontalKB;
    }

    /**
     * Gets the vertical KB.
     * @return - The vertical KB.
     */
    public float getVerticalKB()
    {
        return this.verticalKB;
    }

    /**
     * Gets the attack delay.
     * @return - The attack delay.
     */
    public int getAttackDelay()
    {
        return this.attackDelay;
    }

    /**
     * Updates the knockback values accordingly.
     * @param key - The key of the value you want to update.
     * @param value - The value to update.
     */
    public void update(String key, Object value)
    {
        switch(key)
        {
            case HORIZONTAL_KB:
                this.horizontalKB = (float)value;
                break;
            case VERTICAL_KB:
                this.verticalKB = (float)value;
                break;
            case SPEED_KB:
                this.attackDelay = (int)value;
                break;
        }
    }

    /**
     * Exports the value to a HashMap.
     * @return The hashmap containing the data.
     */
    @Override
    public HashMap<String, Object> export()
    {
        HashMap<String, Object> output = new HashMap<>();
        output.put(HORIZONTAL_KB, this.horizontalKB);
        output.put(VERTICAL_KB, this.verticalKB);
        output.put(SPEED_KB, this.attackDelay);
        return output;
    }

    /**
     * Decodes the object and turns it into a PVPCKnockback instance.
     * @param object - The input object value.
     * @return PvPCKnockback instance or a null value.
     */
    public static PvPCKnockback decode(Object object)
    {
        if(
                object instanceof Map
                && ((Map) object).containsKey(HORIZONTAL_KB)
                && ((Map) object).containsKey(VERTICAL_KB)
                && ((Map) object).containsKey(SPEED_KB)
        )
        {
            Object horizontalKB = ((Map) object).get(HORIZONTAL_KB);
            Object verticalKB = ((Map) object).get(VERTICAL_KB);
            Object speedKB = ((Map) object).get(SPEED_KB);

            // Checks whether or not the PvP Knockback value could be decoded.
            if(
                    horizontalKB instanceof Number
                    && verticalKB instanceof Number
                    && speedKB instanceof Number
            )
            {
                return new PvPCKnockback(
                        ((Number) horizontalKB).floatValue(),
                        ((Number) verticalKB).floatValue(),
                        ((Number) speedKB).intValue()
                );
            }
        }

        return null;
    }
}
